### ISSUE TITLE
Multisite support

### ISSUE THREAD

**Issue by lexislav** [verbb/formie#84]
Support for multisites would be great - are there any plans for that?
- translating form fields
- adjusting form fields settings for differen sites/languages (phone field/phone format, adress field default values etc.)
- editor's premissions to access the forms (based on Sites/SItes group etc.)

**Comment by engram-design** [comment]
It's something earmarked for further down the track. There's a _lot_ to consider, and we're not even sure if it's possible without a significant amount of work. Namely, because form fields inherit from Craft's own fields, which aren't managed on a per-site basis. We'd need to roll out own `fields_sites` table, which would get pretty hairy with complex fields like repeater. 

Then there's the complication of storing the content on a per-site basis. As each form in each site could have different fields. We'd probably have to just have the field settings site-editable.

Similarly, with editors permissions might be difficult as well. Form's are elements, so it'd be like setting up permissions for a specific entry, which isn't really possible in core Craft.

**Comment by giorgiopogliani** [comment]
@engram-design What about having only the submissions multi-site aware? The submissions are just elements that can be multi-site aware, am I wrong? I'm trying to add this because I would need to view and filter the submissions by the site where the form was sent.

**Comment by engram-design** [comment]
That's slightly more achievable, and would also be good to store the site a submission was made on, for a few other issues. There's currently no way to know which site a submission was made on, short of adding custom hidden fields.

I might recommend creating a new issue for multi-site submissions.

**Comment by MattWilcox** [comment]
Just adding a +1 to this. At the moment we're advising the client to create the "same form" twice, with the only difference being the language used for each of them, and telling them to then pick the correct langauge version to feature on a page via the Formie field-type. 

It works, it's just doubling the work for the client each form.

**Comment by engram-design** [comment]
@MattWilcox Yep, I know it's far from ideal, but that'll be the solution for the near future.

**Comment by rtrudel** [comment]
Workaround (and cheap) suggestion: What if the fields rendering use `whateverTheLabelIs|t`. With this set, it could at least translate the default labels (ex: First Name, Last Name, etc.) using Craft translation (or a translation plugin). Or to make it optional (Use translated markup for labels, placeholders and info on rendering).

This is also far from ideal, but at least, most common cases will be handled.

**Comment by engram-design** [comment]
Yep, that's what we currently do, we use `| t` (with no category) which uses the site translations (your `site.php` static translation file). There's really only a small handful of settings that use the `formie.php` static translations class, but they're mostly for control panel based settings.

The issue I believe is mostly around having forms and their fields managed per-site in the control panel.

**Comment by rtrudel** [comment]
This may be a long shot, but this workaround achieves something almost satisfactory. 
You can render the form (using `renderForm()` or another way) after performing some settings overwrites like this:

```twig
    {% set form = craft.formie.forms({ handle: 'contactForm' }).one() %}
    
    {% do form.setFieldSettings('yourName', {        
        firstNamePlaceholder: 'Prénom'|t,
        firstNameErrorMessage: 'Ce champ est requis'|t,
        lastNamePlaceholder: 'Nom de famille'|t,
        lastNameErrorMessage: 'Ce champ est requis'|t,
    }) %}

    {% do form.setFieldSettings('emailAddress', {        
        placeholder: 'Adresse courriel'|t,
        errorMessage: 'Requis et doit être valide'|t,
    }) %}

     {% do form.setFieldSettings('yourPhoneNumber', {        
        placeholder: 'Votre numéro de téléphone'|t,
        errorMessage: 'Requis et doit être valide'|t,
    }) %}

    {% do form.setFieldSettings('subject', {        
        placeholder: 'Sujet du message'|t,
        errorMessage: 'Ce champ est requis'|t,
    }) %}

    {% do form.setFieldSettings('message', {        
        placeholder: 'Votre message'|t,
        errorMessage: 'Ce champ est requis'|t,
    }) %}

    {# RENDER #}
    {{ craft.formie.renderForm(form) }}
```

Then you can use Craft translation system to do the rest. For from perfect, somewhat ugly hack, but working.
   
see https://verbb.io/craft-plugins/formie/docs/developers/field for available attributes.

**Comment by wolfwohlwend** [comment]
+1 An easy way to localize forms in a multisite environment would be phantastic. 99% of our webistes are multilingual and this would make our work significantly easier.

**Comment by wallacio** [comment]
Slightly different requirement, but along the same lines - I'm building a multisite although all in the same language. 

The "Contact" form is common to all sites, the only difference being that the form on each site posts to a different Campaign Monitor list ID. I can appreciate the complexity in not providing that as an option as things currently are but if the form (currently generated from a Stencil) could account for a paramaterised Campaign Monitor list ID, that would be really useful.

Unless I'm being an idiot, the only alternative I can see is to generate a form per site, customising the Campaign Monitor settings for each.

**Comment by engram-design** [comment]
@wallacio That'd be similar to https://github.com/verbb/formie/issues/419 I'd say!

**Comment by wallacio** [comment]
> @wallacio That'd be similar to #419 I'd say!

I'd say you'd be quite correct, and apologies for sloppy searching!! I've made the client realise that actually, a single list will do with a "site" field which they can easily segment it by.

**Comment by JshGrn** [comment]
> Yep, that's what we currently do, we use `| t` (with no category) which uses the site translations (your `site.php` static translation file). There's really only a small handful of settings that use the `formie.php` static translations class, but they're mostly for control panel based settings.
> 
> The issue I believe is mostly around having forms and their fields managed per-site in the control panel.

If I have a 'Name' field with split names using First Name + Last Name, how do I target these in my site.php file? I have tried many combinations but can't see what its trying to match for placeholder and labels? Did anyone else have success with this?

**Comment by engram-design** [comment]
So that will have changed since, as Formie 2 uses the `formie.php` translation category, so everything Formie-translatable will be in that file.

**Comment by JshGrn** [comment]
Ah perfect, must have missed that in the docs.. works perfectly :).

**Comment by daltonrooney** [comment]
I have a translated site and wanted to use the same form for both languages. Kind of building on @rtrudel's example above, I'm working on a small module that accepts a form element, iterates through the fields and matches labels with static translations, updates the field settings, and returns the modified form. This seems to be working fine with my site.php translation file: 

```php 
<?php 

namespace modules\common;
use verbb\formie\fields\formfields\Name;
use verbb\formie\fields\formfields\Address;
use verbb\formie\fields\formfields\Agree;
use verbb\formie\helpers\RichTextHelper;
use Craft;

class FormUtils {
  
  static public function TranslateLabels($form) {
    $fields = $form->getCustomFields();
    foreach ($fields as $field) {
      $form->setFieldSettings($field->handle, [        
        'name' => Craft::t('site', $field->name)
      ]);
      if ($field instanceof Name) {
        $form->setFieldSettings($field->handle, [        
          'prefixLabel' => Craft::t('site', $field->prefixLabel),
          'firstNameLabel' => Craft::t('site', $field->firstNameLabel),
          'middleNameLabel' => Craft::t('site', $field->middleNameLabel),
          'lastNameLabel' => Craft::t('site', $field->lastNameLabel)
        ]);
      }
      if ($field instanceof Address) {
        $form->setFieldSettings($field->handle, [        
          'address1Label' => Craft::t('site', $field->address1Label),
          'address2Label' => Craft::t('site', $field->address2Label),
          'address3Label' => Craft::t('site', $field->address3Label),
          'cityLabel' => Craft::t('site', $field->cityLabel),
          'stateLabel' => Craft::t('site', $field->stateLabel),
          'zipLabel' => Craft::t('site', $field->zipLabel),
          'countryLabel' => Craft::t('site', $field->countryLabel),
        ]);
      }

      if ($field instanceof Agree) {
        $field->setDescriptionHtml($content = Craft::t('site', RichTextHelper::getHtmlContent($field->description)));
      }
    }

    /* More field types here */

    return $form;
  }
}
```

**Comment by jornwildenbeest** [comment]
@engram-design any updates regarding the multisite support?

Will there be an easy way to translate form fields, without adding static strings in the formie.php file, because that defeats the whole purpose of the formie plugin for some of our clients.

**Comment by engram-design** [comment]
@jornwildenbeest No update, it's a massive task, but some of the prep work has been started in Formie 3. You can expect it in Formie 4, which is a long way away yet.

But of course that's the plan, static translations is entirely a stop-gap for the moment.

**Comment by samuelreichor** [comment]
@engram-design what are your thoughts on a setting where we can select a default language for forms? 

At the moment, the craft user language that you can set in the user settings is the default language for forms, which makes total sense, but not for texts that are output in the frontend, I think. 

For example, I always have English as the user language but maintain forms for a German site. So the default error messages and submission texts are in English instead of German. 

This could be avoided by setting a default language for all translations that are displayed in the frontend.

**Comment by jornwildenbeest** [comment]
Hey @engram-design, in #1812 you mentioned "With multisite support also coming".  Do you have an indication of when this feature will be released?

**Comment by engram-design** [comment]
We're working towards it, but as mentioned you can expect it in Formie 4, which is a long way away yet. Likely to come out the same time as Craft 6.

**Comment by engram-design** [comment]
Added in [4.0.0-beta.8](https://github.com/verbb/formie/releases/tag/4.0.0-beta.8).

**Comment by MoritzLost** [comment]
@engram-design Sounds promising! Is there any documentation yet on how multisite support will work in Formie 4? How forms are assigned to sites (can we use propagation methods?), how fields can be translated, etc?

**Comment by engram-design** [comment]
Absolutely! Not all docs are on the website yet but the [multi site](https://github.com/verbb/formie/blob/beta/docs/forms/multi-site.md) docs here on Github is a good start!

But basically most field settings will be translatable by default, just switch the site for your form, and change labels, messages, text, etc. Pretty much anything user-facing can change per site now.

As for propagation, that's possible only with Form Groups (essentially like Sections for entries) where you can define what sites your form should be enabled in, and how propagation works.

<img width="860" height="401" alt="Image" src="https://github.com/user-attachments/assets/5cb6ca38-0118-4278-9226-a60920bfcb7f" />

I think the biggest caveat with multi-site support is that if you want radically different forms (different structure, different fields, radically different settings) then you really are better placed to create different forms, as you have been doing now. It's probably not as super-duper powerful as entries are in that their content can be radically different, but it's close enough I think. But happy to get feedback!

**Comment by MoritzLost** [comment]
@engram-design Thanks, that looks very promising!

I like that you don't create full copies of the form of the site, but only store the differences to the primary form for the localized versions. I also like that only actual text content can be translated, not structural settings. Was worried that a form that exists in many sites might be tedious to maintain, but that makes it easier. I can see some sides needing the ability to override more settings per site – but at that point, you can just create a separate form and maybe use synched fields.

> As for propagation, that's possible only with Form Groups (essentially like Sections for entries) where you can define what sites your form should be enabled in, and how propagation works.

Form groups will be tracked in the project config, right? So we can configure those in dev and the settings will be consistent in all environments?

> I think the biggest caveat with multi-site support is that if you want radically different forms (different structure, different fields, radically different settings) then you really are better placed to create different forms, as you have been doing now. It's probably not as super-duper powerful as entries are in that their content can be radically different, but it's close enough I think. But happy to get feedback!

I'm confused by the use of "primary". Does that refer to the primary site in Craft, or the site a form was created in?
According to the docs, the form is always tied to the primary site:

> Formie stores one canonical copy of each form’s structure and content, tied to Craft’s primary site. 

How does that work when the primary site isn't enabled for a form group?

Those two options are also tripping me up:

- Same language as primary site
- Same site group as primary site

What if the form doesn't exist in the primary site? What if I have a form group that has multiple sites enabled, but _not_ the primary site?

We have some projects where we have multiple site groups (for example, for different regions) and multiple sites in each group (for example, for different languages for the given region). I would expect to be able to set up form groups to mirror those site groups, and have forms propagate either based on language or based on the site group. But the wording above makes it sound like I can _only_ do that for the group with the primary site in it?

**Comment by engram-design** [comment]
> Form groups will be tracked in the project config, right? So we can configure those in dev and the settings will be consistent in all environments?

Correct!

> I'm confused by the use of "primary". Does that refer to the primary site in Craft, or the site a form was created in?
According to the docs, the form is always tied to the primary site:

There are two different ideas here, and the docs could be clearer:

1. Craft’s primary site — the site marked primary under Settings → Sites. This is what we mean when we say canonical content is "tied to the primary site."
2. The site a form was created on — only relevant for the Created site only propagation mode.

So primary is not "wherever you happened to create the form," except in that one propagation mode.

> How does that work when the primary site isn't enabled for a form group?

> What if the form doesn't exist in the primary site? What if I have a form group that has multiple sites enabled, but not the primary site?

Yep, good call, we'll do some improvements here. 

We're likely going to change things:

Mode | Config key | Behaviour
-- | -- | --
All enabled sites | allEnabled | Unchanged. Form enabled on every site in the group’s Enabled Sites list.
Created site only | createdSiteOnly | Unchanged. Form exists only on the site it was created on (must be in enabled list).
Same language as source site | sameLanguage | From enabled sites, keep only those whose language matches this form’s source site.
Same site group as source site | sameSiteGroup | From enabled sites, keep only those in the same Craft site group as this form’s source site.

We're probably also going to switch from "primary" to "source" site — the site where the form was created. Just as a terminology change so people aren't confused with "primary" from Craft site settings.

**Comment by MoritzLost** [comment]
> We're probably also going to switch from "primary" to "source" site — the site where the form was created. Just as a terminology change so people aren't confused with "primary" from Craft site settings.

@engram-design Yes, those changes sound very good! And yeah, using _primary_ might be confusing if it's not talking about the primary Craft site.

FWIW now those options really do seem to mirror the section propagation settings, so it might be a good idea to use the same labels to help people make that connection.
